src/super_gradients/training/utils/callbacks/base_callbacks.py

```diff
@@ -0,0 +1,494 @@
+    TRAIN_EPOCH_START = "TRAIN_EPOCH_START"
+    TRAIN_BATCH_END = "TRAIN_BATCH_END"  # This event corresponds to Callback.on_train_batch_loss_end
+    TRAIN_BATCH_STEP = "TRAIN_BATCH_STEP"  # This event corresponds to Callback.on_train_batch_gradient_step_end
+    TRAIN_EPOCH_END = "TRAIN_EPOCH_END"
+    VALIDATION_BATCH_END = "VALIDATION_BATCH_END"
+    VALIDATION_EPOCH_END = "VALIDATION_EPOCH_END"
+    VALIDATION_END_BEST_EPOCH = "VALIDATION_END_BEST_EPOCH"
+    TEST_BATCH_END = "TEST_BATCH_END"
+    TEST_END = "TEST_END"
+    POST_TRAINING = "POST_TRAINING"
+
+
+class PhaseContext:
+    """
+    Represents the input for phase callbacks, and is constantly updated after callback calls.
+
+    """
+
+    def __init__(
+        self,
+        epoch=None,
+        batch_idx=None,
+        optimizer=None,
+        metrics_dict=None,
+        inputs=None,
+        preds=None,
+        target=None,
+        metrics_compute_fn=None,
+        loss_avg_meter=None,
+        loss_log_items=None,
+        criterion=None,
+        device=None,
+        experiment_name=None,
+        ckpt_dir=None,
+        net=None,
+        lr_warmup_epochs=None,
+        sg_logger=None,
+        train_loader=None,
+        valid_loader=None,
+        training_params=None,
+        ddp_silent_mode=None,
+        checkpoint_params=None,
+        architecture=None,
+        arch_params=None,
+        metric_idx_in_results_tuple=None,
+        metric_to_watch=None,
+        valid_metrics=None,
+        context_methods=None,
+        ema_model=None,
+    ):
+        self.epoch = epoch
+        self.batch_idx = batch_idx
+        self.optimizer = optimizer
+        self.inputs = inputs
+        self.preds = preds
+        self.target = target
+        self.metrics_dict = metrics_dict
+        self.metrics_compute_fn = metrics_compute_fn
+        self.loss_avg_meter = loss_avg_meter
+        self.loss_log_items = loss_log_items
+        self.criterion = criterion
+        self.device = device
+        self.stop_training = False
+        self.experiment_name = experiment_name
+        self.ckpt_dir = ckpt_dir
+        self.net = net
+        self.lr_warmup_epochs = lr_warmup_epochs
+        self.sg_logger = sg_logger
+        self.train_loader = train_loader
+        self.valid_loader = valid_loader
+        self.training_params = training_params
+        self.ddp_silent_mode = ddp_silent_mode
+        self.checkpoint_params = checkpoint_params
+        self.architecture = architecture
+        self.arch_params = arch_params
+        self.metric_to_watch = metric_to_watch
+        self.valid_metrics = valid_metrics
+        self.context_methods = context_methods
+        self.ema_model = ema_model
+
+    def update_context(self, **kwargs):
+        for attr, attr_val in kwargs.items():
+            setattr(self, attr, attr_val)
+
+
+class Callback:
+    """
+    Base callback class with all the callback methods. Derived classes may override one or many of the available events
```

Comment:
Great work on the docs here (:
I'd also add somewhere (perhaps as a comment, like you did in lines 10,11) a mapping for the users who are familiar with the current design but would want to migrate.

Reply:
Totally agree, great suggestion. Added these